### PR TITLE
chore: add generic queue for drop-in actor inbox replacement

### PR DESCRIPTION
**Pull request by stoksc**
## Description
Add a generic queue as a drop-in replacement for an actor inboxes when
removing the actor system. This is useful to avoid side affects where switching
to a channel would introduce new back-pressure mechanisms. In some cases,
it is difficult to vet such back-pressure for unintended side effects. In other cases,
it is impossible to add without a larger change than is wise in a single change.

## Test Plan
- [x] added tests

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/pkg/syncx/queue/queue.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
+package queue
+
+import (
+	"sync"
+)
+
+// Queue is a thread-safe queue.
+type Queue[T any] struct {
+	maxSize int // 0 means no limit
```
Comment (erikwilson) [inline master/pkg/syncx/queue/queue.go:9]: it looks like this variable is not used?

Comment (stoksc) [inline master/pkg/syncx/queue/queue.go:9]:
```suggestion
```

File: master/pkg/syncx/queue/queue.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+	if q.empty() {
+		q.cond.Broadcast()
+	}
+	q.elems = append(q.elems, t)
+}
+
+// Get removes and returns an element from the queue. If the queue is empty, then Get will block
+// until an element is available.
+func (q *Queue[T]) Get() T {
```
Comment (maxrussell) [inline master/pkg/syncx/queue/queue.go:35]: While it's a longer name, I have a mild preference for using the idiomatic operation name of `Dequeue`. Some queues have a "peek" operation and it'd be nice if our naming could save users of the queue a trip to the docs by disambiguating which operation this is.

**Review by maxrussell — CHANGES_REQUESTED** [review]

File: master/pkg/syncx/queue/queue.go (hunk 2 above)
Comment (stoksc) [reply master/pkg/syncx/queue/queue.go:35]: I just replicated the python queue api

File: master/pkg/syncx/queue/queue.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package queue
+
+import (
+	"sync"
+)
+
+// Queue is a thread-safe queue.
+type Queue[T any] struct {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/pkg/syncx/queue/queue.go:8]: Ditto, it is just the same as python. blocking queue is overkill to me - the default type of queue in my mind is blocking.

[the remark this answers is not in the document]
Comment (maxrussell) [reply master/pkg/syncx/queue/queue.go:8]: Huh, that's interesting that we have such opposite views/experiences with queues. [C#'s Queue also doesn't block when the queue is empty (and instead throws an exception)](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.queue-1.dequeue?view=net-7.0#system-collections-generic-queue-1-dequeue). Python's is the only language implementation I can remember ever seeing that defaults to blocking.

Either way, I'm fine with replicating the Python API, including keeping the `Queue` name.

**Comment by stoksc** [comment]
turns out i want this for allocation.go

**Review by maxrussell — APPROVED** [review]

**Comment by stoksc** [comment]
going to merge over unrelated test stuff, since this just adds new, currently unused go files and the go tests pass

**Pull request merged**